File: web/src/core/feature/exchange/data/repository/exchange-offline-first.repository.ts
import type { CupExchange } from "../../domain/entity/CupExchange";
import type {ExchangeRepository} from "../../domain/repository/ExchangeRepository";
import type {ExchangeNetRepository} from "./exchange.net.repository";
import {db} from "../../../../infrastructure/di/dexie.db";
import { toDomain } from "../mapper/Mappers";

const TODAY_CACHE_ID = "directorioCubano-today";

export class ExchangeOfflineFirstRepository implements ExchangeRepository {
    constructor(private readonly net: ExchangeNetRepository) {}

    async getToADay(date: Date): Promise<CupExchange> {
        const cacheId = this.toDayCacheId(date);
        try {
            const remote = toDomain(await this.net.getExchangeToADay(date));
            await db.exchangeRates.put({ ...remote, id: cacheId });
            return remote;
        } catch (error) {
            const cached = await db.exchangeRates.get(cacheId);
            if (cached) return cached;
            throw error;
        }
    }

    async getToARankOfDays(dateFrom: Date, dateTo: Date): Promise<CupExchange[]> {
        try {
            const remote = (await this.net.getExchangeToRankOfDay(dateFrom, dateTo)).map((dto) => toDomain(dto));
            await db.exchangeRates.bulkPut(remote);
            return remote;
        } catch (error) {
            const cached = await db.exchangeRates.toArray();
            if (cached.length) return cached;
            throw error;
        }
    }

    async getCachedToday(): Promise<CupExchange | null> {
        return (await db.exchangeRates.get(TODAY_CACHE_ID)) ?? null;
    }

    async getToday(): Promise<CupExchange> {
        try {
            const remote = toDomain(await this.net.getExchangeToday());
            await db.exchangeRates.put({ ...remote, id: TODAY_CACHE_ID });
            return remote;
        } catch (error) {
            const cached = await db.exchangeRates.get(TODAY_CACHE_ID);
            if (cached) return cached;
            throw error;
        }
    }

    private toDayCacheId(date: Date): string {
        const year = date.getFullYear();
        const month = String(date.getMonth() + 1).padStart(2, "0");
        const day = String(date.getDate()).padStart(2, "0");
        return `directorioCubano-${year}-${month}-${day}`;
    }
}

File: app/src/main/java/com/elitec/alejotaller/feature/exchange/data/repository/ExchangeOfflineFirstRepository.kt
package com.elitec.alejotaller.feature.exchange.data.repository

import com.elitec.alejotaller.feature.exchange.data.dao.ExchangeDao
import com.elitec.alejotaller.feature.exchange.data.mapper.toDomain
import com.elitec.alejotaller.feature.exchange.data.mapper.toLocalDto
import com.elitec.alejotaller.feature.exchange.domain.entity.CupExchange
import com.elitec.alejotaller.feature.exchange.domain.repository.ExchangeRepository

class ExchangeOfflineFirstRepository(
    private val net: ExchangeNetRepository,
    private val dao: ExchangeDao
) : ExchangeRepository {

    private val TODAY_CACHE_ID = "directorioCubano-today"

    override suspend fun getToday(): CupExchange {
        return try {
            val remoteDto = net.getExchangeToday()
            val domain = remoteDto.toDomain()
            // In Web they use TODAY_CACHE_ID as id for the daily one in some places, 
            // but the domain id is derived from date. Let's follow Web's getToday logic:
            // await db.exchangeRates.put({ ...remote, id: TODAY_CACHE_ID });
            dao.insertExchange(domain.copy(id = TODAY_CACHE_ID).toLocalDto())
            domain
        } catch (e: Exception) {
            dao.getExchangeById(TODAY_CACHE_ID)?.toDomain() ?: throw e
        }
    }

    override suspend fun getCachedToday(): CupExchange? {
        return dao.getExchangeById(TODAY_CACHE_ID)?.toDomain()
    }
}
